### PR TITLE
Disable xinput devices when display is blanked

### PR DESCRIPTION
**Pull request by IMbackK**
also Split x11 calls into seperate file

note the todo comment.

### CODE REVIEW

**Comment by IMbackK** [comment]
lets not merge this rn, it causes some wierdness sometimes on resuming the display.

**Comment by IMbackK** [comment]
ok it works reliably now, but its a hack.
turns out x prevents dpms from being disabled while dpms is disabeling the display, this is sensible but polling DPMSInfo for the dpms level and waiting for DPMSModeOff is not enough :(. so there is a sleep(1) in there now :(. idk how to fix this.

maybe polling drm would work, but that would add a huge dependancy

File: modules/display.c [hunk 1]
```diff
@@ -414,7 +381,7 @@ static gboolean brightness_fade_timeout_cb(gpointer data)
 
 	if ((cached_brightness <= 0) && (target_brightness != 0)) {
 		backlight_ioctl(FB_BLANK_UNBLANK);
-		dpms_set_display_on(TRUE);
```
Comment (MerlijnWajer) [inline modules/display.c:417]: I'd like to keep the `dpms_set_display_on` name, at least semantically, explaining that it sets the state on. We went over this at length on the initial PR for the DPMS code. ;)

**Comment by IMbackK** [comment]
@MerlijnWajer please also check this patch with the droid on the lab supply, since x might do something detrimental to pm with the drm interface when disabeling dpms

**Comment by MerlijnWajer** [comment]
I tested this, as noted in IRC, it doesn't seem to idle for me either. I think disabling DPMS probably makes X turn the screen back on again, or something?

```
22:56 < Wizzup> uvos: so I tested your mce, and indeed, the power usage stays high
22:56 < Wizzup> uvos: try 'xset dpms force off' and 'xset -dpms'
22:56 < Wizzup> the power will go back up after the second call
```

One other thing we can consider, which is probably the most ugly hack ever, is to use ptrace to force X to turn off the display after dpms. I wrote a library to easily do these kinds of things years ago:

https://github.com/MerlijnWajer/tracy

**Comment by MerlijnWajer** [comment]
`hw/xfree86/drivers/modesetting/vblank.c` might be of interest, too.

**Comment by IMbackK** [comment]
> I tested this, as noted in IRC, it doesn't seem to idle for me either. I think disabling DPMS probably makes X turn the screen back on again, or something?

I did look and DRM reports the display off, but ill check again. So idk where the power draw comes from.
 
> One other thing we can consider, which is probably the most ugly hack ever, is to use ptrace to force X to turn off the display after dpms. I wrote a library to easily do these kinds of things years ago:
> 
> https://github.com/MerlijnWajer/tracy

I dont like this idea, i would rather patch x than resort to this hack.

File: tklock.c [hunk 2]
```diff
@@ -430,9 +347,6 @@ static gboolean ts_event_control(gboolean enable)
 	execute_datapipe(&touchscreen_suspend_pipe, GINT_TO_POINTER(!enable),
 			USE_INDATA, CACHE_INDATA);
 
-	if (generic_x11_event_control(enable))
-		return TRUE;
-
```
Comment (freemangordon) [inline tklock.c:435]: Why is that chunk removed in a commit named "Split x11 calls into seperate file"? Commit name implies there are no functional changes. Please move that change to a separate commit, along with a relevant commit message.

File: x11-utils.c [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+#include "mce-log.h"
+
+static Atom x11_atom_touchscreen = None;
+static Atom x11_atom_device_enabled = None;
+static Atom x11_atom_device_enabled_type = None;
+static int x11_atom_device_enabled_format = 0;
+
+static Display* x11_get_display(void)
+{
+		Display* dpy = XOpenDisplay(NULL);
```
Comment (freemangordon) [inline x11-utils.c]: Indentation

File: x11-utils.c [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	if (x11_atom_device_enabled == None) {
+		mce_log(LL_WARN, "%s: unable to obtain X11 Atoms", __func__);
+		return false;
+	}
+	
+	if ((x11_atom_device_enabled_type == None) || (x11_atom_device_enabled_format == 0)) {
+		unsigned long ignore_bytes_after, ignore_nitems;
+		unsigned char *ignore_data = NULL;
+
+		if (XIGetProperty(dpy, devinfo.deviceid, x11_atom_device_enabled, 0, 0, False, AnyPropertyType, &x11_atom_device_enabled_type, &x11_atom_device_enabled_format, &ignore_nitems, &ignore_bytes_after, &ignore_data)) {
```
Comment (freemangordon) [inline x11-utils.c]: Please format this to fit some sane number of characters

File: x11-utils.c [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+			mce_log(LL_WARN, "%s: unable to obtain X11 Device Enabled property atom type", __func__);
+			return false;
+		} else {
+			XFree(ignore_data);
+		}
+	}
+	
+	XIChangeProperty(dpy, devinfo.deviceid, x11_atom_device_enabled,
+			x11_atom_device_enabled_type,
+			x11_atom_device_enabled_format, PropModeReplace, (unsigned char*)&enable, 1);
```
Comment (freemangordon) [inline x11-utils.c]: unsigned char *

File: x11-utils.c [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	return true;
+}
+
+bool x11_set_all_input_devices_enabled(Display* dpy, const bool enable)
+{
```
Comment (freemangordon) [inline x11-utils.c:65]: Display *dpy. Also - why the 'const' on that bool?

File: x11-utils.c [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+		mce_log(LL_WARN, "%s: this function only enables devices previously disabled by it", __func__);
+		goto doneXiFree;
+	} else if (!enable && disabledDevices != NULL) {
+		mce_log(LL_WARN, "%s: this function can only disable devices once before renableing them", __func__);
+		goto doneXiFree;
+	} else if (!enable && disabledDevices == NULL) {
+		disabledDevices = malloc(sizeof(*disabledDevices)*ndev);
+	}
+
+	if (!enable) for (int i = 0; i < ndev; ++i) {
```
Comment (freemangordon) [inline x11-utils.c]: please, don;t do that, my eyes bleed :)

File: x11-utils.c [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+		if (devinfo[i].use == XIMasterPointer || devinfo[i].use == XIMasterKeyboard || !devinfo[i].enabled)
+			continue;
```
Comment (freemangordon) [inline x11-utils.c]: indentation

File: x11-utils.c [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+		if (x11_set_input_device_enabled(dpy, devinfo[i], enable)) {
+			disabledDevices[disabledDevicesCount] = devinfo[i];
+			++disabledDevicesCount;
+		}
+	} else {
+		for (int i = 0; i < ndev; ++i) {
+			if (devinfo[i].use == XIMasterPointer || devinfo[i].use == XIMasterKeyboard)
+				continue;
+			
+			if(devinfo[i].name && strstr(devinfo[i].name, "XTEST") != NULL)
```
Comment (freemangordon) [inline x11-utils.c]: if (...

File: x11-utils.c [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+				continue;
+			
+			for(unsigned int j = 0; j < disabledDevicesCount; ++j) {
```
Comment (freemangordon) [inline x11-utils.c]: ditto

File: x11-utils.c [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+	}
+	else {
+		x11_set_all_input_devices_enabled(dpy, true);
+		x11_set_dpms_display_level(dpy, true);
+	}
+	
+	XCloseDisplay(dpy);
+}
+
+bool x11_set_dpms_display_level(Display* dpy, const bool state)
```
Comment (freemangordon) [inline x11-utils.c]: Display *dpy

File: x11-utils.c [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+			XSync(dpy, false);
+		}
+	}
+	else {
+		mce_log(LL_WARN, "%s: Dsiplay dose not support DPMS", __func__);
+		if(ownsDisplay)
+			XCloseDisplay(dpy);
+		return false;
+	}
+
```
Comment (freemangordon) [inline x11-utils.c:196]: extra new line

File: x11-utils.h [hunk 13]
```diff
@@ -0,0 +1,18 @@
+#ifndef _X11_UTILS_H_
+#define _X11_UTILS_H_
+#include <stdbool.h>
+#include <X11/extensions/XInput.h>
+#include <X11/extensions/XInput2.h>
+
+bool x11_set_input_device_enabled(Display* dpy, const XIDeviceInfo devinfo, const bool enable);
```
Comment (freemangordon) [inline x11-utils.h:7]: Display *dpy

**Review by freemangordon — CHANGES_REQUESTED** [review]

File: tklock.c [hunk 14]
```diff
@@ -347,8 +347,7 @@ static gboolean ts_event_control(gboolean enable)
 	execute_datapipe(&touchscreen_suspend_pipe, GINT_TO_POINTER(!enable),
 			USE_INDATA, CACHE_INDATA);
 
-	return generic_event_control(mce_touchscreen_sysfs_disable_path,
```
Comment (freemangordon) [inline tklock.c:436]: Please explain in the commit message why is this needed.

File: tklock.c (hunk 2 above)
Comment (IMbackK) [reply tklock.c:435]: Its not called that. The commit is called:  
Disable xinput devices when display is blanked 
Split x11 calls into seperate file

disabeling the touchscreen xinput device is redundant if you also disable all xinput devices.

File: x11-utils.c (hunk 6 above)
Comment (IMbackK) [reply x11-utils.c:65]: Why not const on that bool? we dont intend modify it, this is best practice.

File: tklock.c (hunk 14 above)
Comment (IMbackK) [reply tklock.c:436]: in the old code that function allways returns here: 
https://github.com/maemo-leste/mce/blob/3cbee89d364eea0211fa7a37493e39a6c2b22192/tklock.c#L434

unless the generic_x11_event_control returns false, witch it should never do since that indicates a failure communicating with X11.

because of this generic_event_control never gets called. To preserve the same behavior i had to remove it when removing https://github.com/maemo-leste/mce/blob/3cbee89d364eea0211fa7a37493e39a6c2b22192/tklock.c#L433 (see above as to why that was removed)

Honestly this looks like a possible bug to me, but i dont know and its unrelated to what i was doing anyways so i left it alone.

File: x11-utils.c (hunk 6 above)
Comment (freemangordon) [reply x11-utils.c:65]: I was just wondering how it is better, given that the variable is passed by value anyway.

File: x11-utils.c (hunk 7 above)
Comment (IMbackK) [reply x11-utils.c]: i find this more readable than the alternative but ok its changed

File: x11-utils.c (hunk 6 above)
Comment (IMbackK) [reply x11-utils.c:65]: Well it conveys intention: this variable is not supposed to be modified that allows the compiler to provide better warnings when a future programmer mistypes (classic = vs == typo). It also helps the compiler to optimize away the copy in some instances, but these instances are not relevant here.

**Comment by freemangordon** [comment]
I am fine with it, @MerlijnWajer - shall we merge?

**Comment by MerlijnWajer** [comment]
I'm ok with it - but keep in mind that this breaks volume and other keys being sent to volumeapplet and such, iiuc.

We can merge it now, but we need a better fix, in the form of a X11 patch.

**Comment by IMbackK** [comment]
```
21:05 freemangordon: BTW, I think we shall discuss how to fix that broken "dpms disable" functionality
21:05 uvos: freemangordon: this pr 1. refactos x11 code and 2. disables all input devices on dpms off
21:05 uvos: freemangordon: 2 is a temporary solution only
21:06 freemangordon: right
21:06 uvos: freemangordon: i will remove it as soon as we have something better
21:06 freemangordon: we shall patch X I guess
21:06 uvos: freemangordon: yes maybe or patch hildon to work with all displays disabled in xrandr
21:06 uvos: freemangordon: either way i will ajust mce when the time comes
21:06 freemangordon: ok
21:07 uvos: i have pushed a fix for the confusing brackts
21:07 uvos: *brackets
21:10 freemangordon: ok, lets see if Wizzup has any objections, I guess he'll just merge. If not, I'll do it
```

File: x11-utils.c [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+			DPMSForceLevel(dpy, DPMSModeOn);
+			XSync(dpy, false);
+		} else {
+			usleep(100000);
+			DPMSForceLevel(dpy, DPMSModeOff);
+			XSync(dpy, false);
+		}
+	}
+	else {
+		mce_log(LL_WARN, "%s: Dsiplay dose not support DPMS", __func__);
```
Comment (spinal84) [inline x11-utils.c:191]: dose => does

File: x11-utils.c [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+		if(ownsDisplay)
```
Comment (spinal84) [inline x11-utils.c]: 'if(' => 'if ('

File: x11-utils.c [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+	}
+
+	if (!enable) for (int i = 0; i < ndev; ++i) {
+		if (devinfo[i].use == XIMasterPointer || devinfo[i].use == XIMasterKeyboard || !devinfo[i].enabled)
+			continue;
+		
+		if(devinfo[i].name && strstr(devinfo[i].name, "XTEST") != NULL)
+			continue;
+	
+		mce_log(LL_INFO, "%s: disableing %s", __func__, devinfo[i].name);
```
Comment (spinal84) [inline x11-utils.c]: disableing => disabling

**Comment by IMbackK** [comment]
@spinal84 uh. none of this code is current. its been refactored into a module https://github.com/maemo-leste/mce/blob/master/modules/x11-ctrl.c

**Comment by spinal84** [comment]
> @spinal84 uh. none of this code is current. its been refactored into a module https://github.com/maemo-leste/mce/blob/master/modules/x11-ctrl.c

Good job.
